gramdictru/Scripts/Utils/copyUtils.tsx

```diff
@@ -0,0 +1,273 @@
+﻿import * as React from "react";
+
+export const text2Markdown = (text: string) => {
+    if (!text) {
+        return ""
+    }
+    return (text as any)
+        .replaceAll('_', '\_')
+        .replaceAll('<i>', '_')
+        .replaceAll('</i>', '_')
+        .replaceAll('<b>', '__')
+        .replaceAll('</b>', '__')
+}
+
+class CopyUtils {
+
+    getOriginNode(node) {
+        if (node.nodeName === '#text') {
+            return node.parentNode
+        }
+
+        return node
+    }
+
+    isBold(node) {
+        if (!node) {
+            return false
+        }
+
+        if (node && node.nodeName === '#text') {
+            return node.parentNode.nodeName === 'B' || this.isBold(node.parentElement)
+        }
+
+        let fontWeight = getComputedStyle(node).fontWeight
+        return getComputedStyle(node).fontWeight === "bold" || parseInt(fontWeight) > 400
+    }
+
+    isItalic(node) {
+        if (!node) {
+            return false
+        }
+
+        if (node.nodeName == 'I') {
+            return true
+        }
+
+        if (node.nodeName === '#text') {
+            return node.parentNode.nodeName === 'I' || this.isItalic(node.parentElement)
+        }
+
+        return getComputedStyle(node).fontStyle === "italic"
+    }
+
+    getParentCopyField(node) {
+        if (node.classList && node.classList.contains("copy-field")) {
+            return node
+        }
+
+        return this.getParentCopyField(node.parentElement)
+    }
+
+    getFirstNodeText(selection, anchorNode, focusNode, isReverseCopy) {
+        let text = ''
+        if (isReverseCopy) {
+            if (focusNode.classList.contains("lemma")) {
+                let preText = selection.focusNode.textContent.substring(
+                    selection.focusOffset)
+                text = preText ? `__${preText}__` + " " : ""
+            } else if (focusNode.classList.contains("symbol")) {
+                let preText = selection.focusNode.textContent.substring(
+                    selection.focusOffset,
+                )
+                text = preText ? preText + " " : ""
+            } else if (focusNode.classList.contains("grammar")) {
+                let grammarText = ""
+                var focusIndex = [...focusNode.childNodes].findIndex(_node => _node === selection.focusNode || _node === selection.focusNode.parentNode)
+                for (var [i, _node] of focusNode.childNodes.entries()) {
```

Comment:
Could you please get rid of "Duplicate declaration" for variables `i` and `_node`? (Here and 3 more occurrences below). Also, why does `_node` have an underscore? Does it mean something?

Reply:
> Could you please get rid of "Duplicate declaration" for variables `i` and `_node`? (Here and 3 more occurrences below). Also, why does `_node` have an underscore? Does it mean something?

 I don't know why you get warning "Duplicate declaration", because "i" and "_node" are local variables, and will be clear when function finish. "_node" in line 76 will be clear before the code in line 77 has run

sometime i use the variable has underscore for local variable, but i just remove it in latest code, please check again